langmeta-sqlite/src/main/scala/org/langmeta/sqlite/Main.scala

```diff
@@ -0,0 +1,214 @@
+
+          val semanticdbStart = System.nanoTime()
+          val documentsToPaths = mutable.Map[String, Path]()
+          var documentId = new Counter()
+          var nameId = new Counter()
+          var messageId = new Counter()
+          var _symbolId = new Counter()
+          var symbolIds = mutable.Map[String, Int]()
+          var symbolTodo = mutable.Map[Int, String]()
+          var symbolDone = mutable.Set[Int]()
+          def symbolId(symbol: String): Int = {
+            if (symbolIds.contains(symbol)) {
+              symbolIds(symbol)
+            } else {
+              val symbolId = _symbolId.next()
+              symbolIds(symbol) = symbolId
+              symbolTodo(symbolId) = symbol
+              symbolId
+            }
+          }
+          val syntheticId = new Counter()
+
+          semanticdbFilenames.foreach { semanticdbFilename =>
+            val path = Paths.get(semanticdbFilename)
+            try {
+              val bytes = Files.readAllBytes(path)
+              val db = Database.parseFrom(bytes)
+              db.documents.foreach { document =>
+                documentsToPaths.get(document.filename) match {
+                  case Some(existingPath) =>
+                    val what = document.filename
+                    val details = "both $existingPath and $path"
+                    println(s"Duplicate document filename $what in $details")
+                  case None =>
+                    genuineDocuments += 1
+                    documentsToPaths(document.filename) = path
+
+                    val documentRef = documentId.next
+                    documentStmt.setInt(1, documentRef)
+                    documentStmt.setString(2, document.filename)
+                    documentStmt.setString(3, document.contents)
+                    documentStmt.setString(4, document.language)
+                    documentStmt.executeUpdate()
+
+                    document.names.foreach { name =>
+                      nameStmt.setInt(1, nameId.next)
+                      nameStmt.setInt(2, documentRef)
+                      nameStmt.setInt(3, name.position.get.start)
+                      nameStmt.setInt(4, name.position.get.end)
+                      nameStmt.setInt(5, symbolId(name.symbol))
+                      nameStmt.setBoolean(6, name.isDefinition)
+                      nameStmt.executeUpdate()
+                    }
+
+                    document.messages.foreach { message =>
+                      messageStmt.setInt(1, messageId.next)
+                      messageStmt.setInt(2, documentRef)
+                      messageStmt.setInt(3, message.position.get.start)
+                      messageStmt.setInt(4, message.position.get.end)
+                      messageStmt.setInt(5, message.severity.value)
+                      messageStmt.setString(6, message.text)
+                      messageStmt.executeUpdate()
+                    }
+
+                    document.symbols.foreach { symbol =>
```

Comment:
I propose we turn this into an `Iterator[schema.Document]` and move the logic out of the main function

Reply:
This file definitely needs restructuring. I hacked it up in a couple of hours, so I optimized for writing, not for reading. The results were so exciting that I rushed to share them without any cleanup.